### ISSUE TITLE
Feature request: add a location variable

### ISSUE THREAD

**Issue by flavio20002** [Waboodoo/HTTP-Shortcuts#95]
Hi, 
I would like to add a new kind of variable, that contains the current device position at the moment the shortcut is triggered. My idea is to make it similar to color picker: it could show a little "open map" with the current position and a pin that can be moved on the map. The OK button, extract the position of the pin and send it in the body of the request as a Json. 
I will start to develop it in the next days, if you agree.
Thanks.

Flavio

**Comment by flavio20002** [comment]
I try to answer:

- I would wrap the two variable in a json string: {"latitude":"45.554926","longitude":"13.693004"}. The server should be able to parse this kind of input. In my case, I will use Home Assistant to manually update a position (my car, after I parked)

- For Android API version < 21 the new permission will be requested while install with no choice of the user, after that, we will request runtime and the user will be able to reject it. In . both cases the privacy policy needs to be updated.

- I would put a maximum accuracy parameter (or a "quality level", and a maximum number of attempts. I've done in other Apps with great results. Sometimes, the first Gps fix is not accurate, but usually the second or the third is accurate enough. For example, my car need a very accurate position, otherwise I should have to search it within 150 meters. My workaround now is using an IFFT widget that send my position by email, only after visiting Google Maps and get an accurate enough position. If the App is well studied, we could accomplished it with one finger. If the position is not accurate enough for that quality level, or the position is not granted, an error popup could appear.

**Comment by Waboodoo** [comment]
Ok, I suggest we try it out.

I'm not happy with encoding the variable as a JSON string, because that restricts how it can be used as it makes the assumption that the server expects a JSON object of exactly that structure. As a first prototype, a think we can try it like that, but I'd like to generalize that later on.

I'm imagining an approach where the variable itself gets the latitude and longitude assigned as a single value (e.g. "45.554926,13.693004"), but there would also be 2 additional "virtual" variables that get the latitude and longitude assigned separately. E.g. if the variable itself is called "location", you could use {{location}} to get the combined values of latitude and longitude, but you could also use the virtual variables {{$location.latitude}} and {{$location.longitude}} to get the separate values. This would then allow for using the 2 values more freely in an arbitrary JSON body.

**Comment by flavio20002** [comment]
Almost done. Just some little details and I will try a PR.
I wanted to preview that I'm using this [library](https://github.com/osmdroid/osmdroid) to display the map. What do you think about? Here's a screenshot.

![screenshot_1524429080](https://user-images.githubusercontent.com/10809301/39099542-fa8adf02-467c-11e8-9103-9dc3ad8e917d.png)

**Comment by hadmut** [comment]
That's somewhat overcomplicated. 

A common practice (e.g. on google maps) is to just use the latitude and longitude (and optionally the elevation/altitude) as comma separated floating numbers, sometimes prefixed with a „@“ to mark a location, as a very simple syntax. Just two or three numbers with commas between.  If you want a small, lean implementation, just generate a string like that, especially when choosing a point on a map. 

Unfortunately, there's some MQTT application which use a colon : instead of a comma as a separator, which would take substution on the server side. 


If you want a better, more powerfull implementation, generate a json record containing all what's available from android, like bearing, speed, altitude, acuracy, simply all what can be found in the Android Location object, as described in https://developer.android.com/reference/android/location/Location


Best way would probably be to implement both as two distinct variable types, the simple lat,long string and the json record containing all available data. 

Or, as a single variable type, which lets you choose from options whether you want
- simple, comma separated    LAT,LON
- simple, colon separated    LAT:LON
- optionally with altitude
- json full record with all what's available from OS


regards
Hadmut

**Comment by pataquets** [comment]
There is another notation to express a location: [Plus Codes](https://maps.google.com/pluscodes/).
They're used in Google Maps, but the spec is open source and they look like quite powerful and flexible.
No idea of the required effort to implement them, thou.
Just my 2c.

**Comment by sebastiannielsen** [comment]
@flavio20002 I would also wish to be able to specify that it should immidiately send the location without the map popup. For usage for example as duress alarm (Send my location to my server which will then echo it out to all my friends). It isn't exactly a good idea to have to OK a map dialog in a violent situation or similiar.

**Comment by Waboodoo** [comment]
@sebastiannielsen Thanks for the request. First of all, just a small but important disclaimer: please don't rely on the app to perform life or mission critical tasks. It is not designed for that and I can not give any guarantees, so, use at your own risk.

Regarding your request: I will have to think about how to do this. Originally, my main concern for such a feature was that it would mean that the app requires the Location permission, which can make the app look a bit fishy, given that it's not immediately clear to a user why an HTTP app would need to know their location. But in the meantime, since the last time this topic came up, I had to add the Location permission to the app anyway, because it's needed to request the current wifi SSID, and that didn't seem to have as bad an impact as I expected, so I suppose that's not a big issue.

My other concern is that a location consists of multiple parts (latitude, longitude, possibly altitude and accuracy), and a value might not always be available. This does not fit so well into the concept of variables, which in their current form can only hold a single textual value.

So, thinking out loud, perhaps it would be better if instead I introduce this as a function for the Scripting feature, e.g. `requestLocation()`, which can take some configuration parameters and return a JS object containing the location and possibly some meta data. This will make it a bit more complex to use, but much more flexible. I'm imagining that you could have a shortcut with a Javascript code snippet like the following:

```js
const location = requestLocation({
    'timeout': 5000,
});
// location is now an object with keys 'latitude' and 'longitude', which can be stored into variables or used in other ways
if (location.latitude != null) {
    setVariable('my_latitude', location.latitude);
    setVariable('my_longitude', location.longitude);
} else {
    alert("No location...");
    abort();
}
```

Do you think this would work for your use case?

I know this makes it less accessible for many users, as it requires at least some basic programming skills as opposed, so it's definitely not perfect. I'm thinking about maybe adding a bridge to allow using the Scripting capabilities without needing to write code.

**Comment by sebastiannielsen** [comment]
@Waboodoo I would prefer a input variable as I described in #274, and then have it being set as string:
The multiple coordinate could be simply made by separating the fields by a comma.

Server can easily parse out this by running:
($latitude, $longitude, $fixlevel) = split(",", param('coordinate'));

So setting variable "coordinate = GPS_input"
will result in:
coordinate=xx.xxxxxxxx,yy.yyyyyyyy,z

x = latitude
y = longitude
where z = 0 for no fix, z = 1 weak fix, and z = 2 full fix

and then of course the setting I suggested where you can specify "Wait for fix timeout".

**Comment by sebastiannielsen** [comment]
@Waboodoo what do you think about my suggestion, separating the values with comma?

I personally think it would be the best solution, as most people who use HTTP Shortcuts also control the server side application, and can do the neccessary splitting.

And in case of not, a simple JS code can be written to split up the values like your idea.

So the default is to send the values separated by commas as one single string, but the users that want to send it to a third party application which does not support the format, then the user can write a pre-exec JS to handle the input and convert it.

This would become great to release along with the QR Code Droid feature, so the locations of scans can be collected too, meaning excellent for delivery drivers to use to register their drop offs :-)

**Comment by Waboodoo** [comment]
> as most people who use HTTP Shortcuts also control the server side application

I don't believe that is actually true. I don't have any real data on how common different use cases are in this app, but I know that there's a lot of them, and in many cases the user is not in control of the server. And even if you are in control of the server, it is likely still easier to add a handful of lines of code on the client than to modify the server. This is one reason why I don't want to add such a specific implementation detail that would only work for certain use cases, and instead favor a general approach.

For convenience, I can still include a pre-concatenated string version of the latitude and longitude in the data object that is returned by the location lookup function, so the resulting object might look something like this:

```js
{
    "coordinates": "56.6745584,-5.0830197",
    "latitude": 56.6745584,
    "longitude": -5.0830197,
    // and potentially additional information such as precision, velocity, bearing, ...
}
```

I have not yet decided on how exactly this will work, I'll do some more discovery.

**Comment by sebastiannielsen** [comment]
@Waboodoo Thats why I suggest to have the location function return a pre-concatenaded string (so you can just choose "GPS Location" as a input type, in variables, and then set fix timeout there, and it will just send the pre-concatenated string to server), but in case you dont have access to the server, you can just have a JS script, that reads the variable, and then replaces it with the splitted data that you really want.

So those that don't want a client side script, can just use the function as-is as input data (and write the script server-side instead), and those that want to write a client side JS script, can do that.

Its just, that I don't want a forced JS approach, where a client-side JS-script is REQUIRED to use the location function.

**Comment by Waboodoo** [comment]
I see your point, and I agree that it adds additional complexity to setting up the app and makes it less accessible for those who are not familiar with JS. This touches on a bigger topic that I've been thinking about for a long time now, essentially the question of Scripting vs. Variables. This question arises for any kind of data that needs to be queried: "should this be accessible through a variable or should it be a JS function?". Currently my stance is that for all of these highly specific and rarely used features (which querying the location definitely is one of) going the Scripting route is better because it allows for more flexibility and doesn't clutter up the app as much in terms of UI. These features are primarily used by power users, so we can allow for a bit of extra complexity and effort in setting something up.

I have a vision that eventually the app will offer a way to better integrate variables with Scripting and vice-versa, so that it will be possible to use some of these features without needing to write any JS code, but I have not quite figured out how to go there, and I don't want to open that can of worms as part of this feature now. But it will give me some food for thought, so let's call it an experiment.

What I definitely don't want to do though is to invent new data encoding formats that are not human-readable, such as appending a mysterious number, where you have to look up the documentation to know what it might mean. So encoding the status (e.g. accuracy) of the location result will have to be done differently.

**Comment by sebastiannielsen** [comment]
@Waboodoo Agreed, Thats why I suggest making all functions that DO NOT require an input (like location or QR code - it doesn't require external input data) an available variable function.

Power users can STILL use the function with JS, by simply reading the variable, processing it, and replacing it with the intended results. (and using additional variables if required)

**Comment by Waboodoo** [comment]
I'll keep the thought, and I made some notes about it in my backlog. There's definitely room for improvement.

For now, I'm closing this issue as support for this feature was added in 2.18.0, which is being released now.
Documentation can be found here: https://http-shortcuts.rmy.ch/scripting#get-location
